File: js/kata_are_they_the_same.js
function comp(a, b) {
    {
        // Preliminary condition tests:
        if (a === b) return true;
        if (!a || !b) return false;
        if (a.length !== b.length) return false;

        // map a to square a, and sort both arrays:
        a = a.map(item => item*item).sort();
        b = b.sort();

        // compare each element of a and b:
        for (let i = 0; i< a.length; i++) {
            if (a[i] !== b[i]) {
                return false;
            }
        }

        return true;
    }
}

let a = [121, 144, 19, 161, 19, 144, 19, 11];
let b = [14641, 20736, 361, 25921, 361, 20736, 361, 121];

let test = comp(a, b);
console.log(`Test 1: ${test}`);

a = [121, 144, 19, 161, 19, 144, 19, 11];
b = [14641, 20736, 361, 25921, 361, 20736, 362, 121];
test = comp(a, b);
console.log(`Test 2: ${test}`);

File: cpp/kata_are_they_the_same.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::cout;
using std::endl;
using std::vector;

class Same
{
public:
    static bool comp(vector<int> &a, vector<int> &b)
    {
        if (a.size() != b.size()) return false;
        for (vector<int>::size_type i = 0; i< a.size(); i++) {
            a[i] = a[i]*a[i];
        }
        std::sort(a.begin(), a.end());
        std::sort(b.begin(), b.end());
        for (vector<int>::size_type i = 0; i< a.size(); i++) {
            if (a[i] != b[i]) {
                return false;
            }
        }

        return true;
    }
};

int main(int, char **)
{
    vector<int> a = {121, 144, 19, 161, 19, 144, 19, 11};
    vector<int> b = {14641, 20736, 361, 25921, 361, 20736, 361, 121};

    bool test = Same::comp(a, b);
    cout << "Test 1: " << test << endl;

    a = {121, 144, 19, 161, 19, 144, 19, 11};
    b = {14641, 20736, 361, 25921, 361, 20736, 362, 121};
    test = Same::comp(a, b);
    cout << "Test 2: " << test << endl;
}
